added system prompt for openai

Leo-LiHao opened the pull request:
# What does this PR do?
Add system prompt for OpenAI Representation model, see Issue #2146 



## Before submitting
- [ ] This PR fixes a typo or improves the docs (if yes, ignore all other checks!).
- [x] Did you read the [contributor guideline](https://github.com/MaartenGr/BERTopic/blob/master/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes (if applicable)?
- [ ] Did you write any new necessary tests?

MaartenGr commented on bertopic/representation/_cohere.py:
```diff
@@ -33,7 +33,9 @@
 [DOCUMENTS]
 Keywords: [KEYWORDS]
-Topic name:"""
+Provide the topic name directly without any explanation."""
```
Why did you make this change? In my experience, by providing the model with a pre-fix, there is no need for mentioning that it should provide the topic name without any explanation.

Leo-LiHao replied:
Based on my experiments, the pre-fix prompt works well for completion models but not with (small) chat models. You can try different prompts locally to see the difference and feel free to change/iterate the prompt.

MaartenGr replied:
Let's keep it the same it was. Updating the prompt would require another PR with additional testing. Also, this changes the way the few-shot prompt was created.

MaartenGr reviewed (commented):
Thanks for updating these two models! I left a small comment. 

Other than that, I believe we will need to look at the following:
* [HuggingFace](https://github.com/MaartenGr/BERTopic/blob/master/bertopic/representation/_textgeneration.py)
* [LangChain](https://github.com/MaartenGr/BERTopic/blob/master/bertopic/representation/_langchain.py)
* [Llama-cpp-python](https://github.com/MaartenGr/BERTopic/blob/master/bertopic/representation/_llamacpp.py)

With HuggingFace, we might need to change the formatting such that we use messages instead:

```python
messages = [
    {
        "role": "system",
        "content": MY_SYSTEM_PROMPT,
    },
    {"role": "user", "content": prompt},
 ]
```

For LangChain, I'm okay with skipping this for now since it uses a more complex structure and the API here might need to be updated (see https://python.langchain.com/v0.1/docs/modules/model_io/chat/quick_start/#messages-in---message-out).

For llama-cpp-python, it seems straightforward and we should use [`.create_chat_completion`](https://github.com/abetlen/llama-cpp-python?tab=readme-ov-file#chat-completion) instead of directly calling the model.

Leo-LiHao commented:
@MaartenGr I updated the code for llama-cpp

MaartenGr commented on bertopic/representation/_llamacpp.py:
```diff
@@ -8,13 +8,34 @@
+Sample texts from this topic:
 [DOCUMENTS]
+Keywords: [KEYWORDS]
+Provide the extracted topic name directly without any explanation."""
```
Let's structure this the same way as was originally done in `_cohere.py`

MaartenGr commented on bertopic/representation/_cohere.py:
```diff
@@ -33,7 +33,9 @@
-Topic name:"""
+Provide the topic name directly without any explanation."""
+
+DEFAULT_SYSTEM_PROMPT = "You are designated as an assistant that identify and extract high-level topics from texts."
```
```suggestion
DEFAULT_SYSTEM_PROMPT = "You are an assistant that extracts high-level topics from texts."
```

MaartenGr commented on bertopic/representation/_llamacpp.py:
```diff
@@ -8,13 +8,34 @@
-
-Based on the above information, can you give a short label of the topic?
-A: """
+DEFAULT_SYSTEM_PROMPT = "You are designated as an assistant that identify and extract high-level topics from texts."
```
```suggestion
DEFAULT_SYSTEM_PROMPT = "You are an assistant that extracts high-level topics from texts."
```

MaartenGr commented on bertopic/representation/_llamacpp.py:
```diff
@@ -93,14 +114,15 @@ def __init__(
         nr_docs: int = 4,
         diversity: float = None,
-        doc_length: int = None,
+        doc_length: int = 100,
```
Let's keep the value as it was, otherwise we have to do this for all instances of `doc_length` across all models.

MaartenGr commented on bertopic/representation/_llamacpp.py:
```diff
@@ -93,14 +114,15 @@ def __init__(
     ):
         if isinstance(model, str):
-            self.model = Llama(model_path=model, n_gpu_layers=-1, stop="Q:")
+            self.model = Llama(model_path=model, n_gpu_layers=-1, stop="\n", chat_format="llama-2")
```
What makes you choose `llama-2` as the default chat template? Should we choose one at all or are that perhaps more common templates like `ChatML`?

Leo-LiHao replied:
I followed the readme in llamacpp, I have changed it to `ChatML`

MaartenGr commented on bertopic/representation/_llamacpp.py:
```diff
@@ -110,6 +132,7 @@ def __init__(
                 "local LLM or a ` llama_cpp.Llama` object."
             )
         self.prompt = prompt if prompt is not None else DEFAULT_PROMPT
+        self.system_prompt = system_prompt if system_prompt is not None else DEFAULT_SYSTEM_PROMPT
```
Should there also be a `self.default_system_prompt_`?

MaartenGr commented on bertopic/representation/_llamacpp.py:
```diff
@@ -150,33 +173,31 @@ def extract_topics(
+                      }
+                  ], ** self.pipeline_kwargs
+            )
+            label = topic_description["choices"][0]["message"]["content"].strip().replace("Topic name: ", "")
```
I think we can remove `.replace("Topic name: ", "")` if the prompt is updated (see comment above).

MaartenGr commented on bertopic/representation/_llamacpp.py:
```diff
@@ -150,33 +173,31 @@ def extract_topics(
-                for doc in docs:
-                    to_replace += f"- {doc}\n"
-                prompt = prompt.replace("[DOCUMENTS]", to_replace)
+        prompt = self.prompt
+        if "[KEYWORDS]" in prompt:
+            prompt = prompt.replace("[KEYWORDS]", keywords)
+        if "[DOCUMENTS]" in prompt:
+            to_replace = ""
+            for doc in docs:
+                to_replace += f"- {doc}\n"
+            prompt = prompt.replace("[DOCUMENTS]", to_replace)
```
If we update this piece of code here, shouldn't we also do this for the other models, like `_cohere.py`?

Leo-LiHao replied:
I think it is a bug in the old implementation which does not replace `[DOCUMENTS]` for `DEFAULT_PROMPT`. I have replaced `_create_prompt` in `_llamacpp.py` as the one in `_cohere.py`

MaartenGr commented on bertopic/representation/_openai.py:
```diff
@@ -48,6 +48,8 @@
 topic: <topic label>
 """
 
+DEFAULT_SYSTEM_PROMPT = "You are designated as an assistant that identify and extract high-level topics from texts."
```
```suggestion
DEFAULT_SYSTEM_PROMPT = "You are an assistant that extracts high-level topics from texts."
```

MaartenGr commented on bertopic/representation/_utils.py:
```diff
@@ -44,8 +44,8 @@ def decode(self, doc_chunks):
     if doc_length is not None:
-        if tokenizer == "char":
-            truncated_document = document[:doc_length]
+        if tokenizer == "char" or tokenizer is None:
+            truncated_document = document[:doc_length] + " (...)"
```
Could you explain why this was added? I believe something like this was addressed in https://github.com/MaartenGr/BERTopic/pull/2181

Leo-LiHao replied:
Yes, some issue which was addressed in #2181

MaartenGr reviewed (commented):
Apologies for the late reply, I was off for the last couple of weeks. 

I left some comments here and there to make sure we are on the same page for all changes.

Leo-LiHao commented:
> Apologies for the late reply, I was off for the last couple of weeks.
> 
> I left some comments here and there to make sure we are on the same page for all changes.

@MaartenGr Thanks for your comments, I incorporated them and updated the code.

MaartenGr commented:
@Leo-LiHao Thanks! I noticed a couple of changes to creating the prompt that we might want in all other models but it's alright to keep that as is for now. At some point, I might make a utility function for creating the prompt that is used across all models to make sure everything remains stable. 

For now, let's merge!

MaartenGr commented:
Ah, I see the linting has an error. Could you check that?

Leo-LiHao commented:
@MaartenGr Thank you! I fixed the lint error and it is read for merge now.

MaartenGr commented:
Awesome, thanks for taking the time! I just merged it 🥳
